import os
import glob
from subprocess import Popen


def run_srst2(fastq, gene_db, run_number):
    srst2_output = fastq.replace('.fastq.gz', '_run{}_SRST2_output'.format(run_number))
    cmd = 'srst2 ' \
          '--input_se {input_se} ' \
          '--gene_db {gene_db} ' \
          '--output {output} ' \
          '--threads {threads} ' \
          ''.format(input_se=fastq, gene_db=gene_db, output=srst2_output, threads=10)
    p = Popen(cmd, shell=True)
    p.wait()


def main():
    gene_db = '/home/forest/PycharmProjects/fetagenome_generator/amr_targets/NCBI_AMR_nt_180312_SRST2.fasta'

    fetagenome_list = [
        '/home/forest/Documents/Projects/Fetagenome_testing/subsampled_data/subsampled_250000/merged_fetagenome.fastq.gz',
    ]

    n_runs = 3

    for fetagenome in fetagenome_list:
        print('Running srst2 on {}'.format(fetagenome))
        for n in range(n_runs):
            run_srst2(fastq=fetagenome, gene_db=gene_db, run_number=n+1)

        # Cleanup the .pileup files generated by srst2. They're big.
        tmpdir = os.path.dirname(fetagenome)
        cleanup = glob.glob(os.path.join(tmpdir, '*.pileup'))
        for pileup in cleanup:
            os.remove(pileup)


if __name__ == '__main__':
    main()
